### DISCUSSION TITLE
Initial delay on cron scheduled trigger

### DISCUSSION THREAD

**Discussion by maximauro** [quartznet/quartznet#1082] [Q&A]
I have a replicated application (let's say 2 replicas) executing a cron scheduled job that read/writes in a database.
To avoid lock problems, I don't want each replica to trigger the job at the exact same time.

Example 1 : My trigger is configured with the cron expression 0 0 * * * (midnight every day)
I want Replica 1 to execute at midnight, and Replica 2 to execute at midnight + given delay (x seconds or minutes).

Example 2 : My trigger is configured with the cron expression */10 * * * * (every 10 minutes)
I want Replica 1 to execute at 00:10, 00:20, 00:30, etc...,
And Replica 2 to execute at 00:10+delay, 00:20+delay, 00:30+delay, etc...,

A possible solution would be to be able to configure a delay on each execution. 
Maybe something like : WithCronSchedule("0 0 * * *").WithDelayInSeconds(x)

An alternative is to manually execute a Thread.Sleep() with a value depending on the replicated instance number...

**Comment by lahma** [comment]
I know that cron expression format is nice configuration-wise, but what you are asking is basically interval based trigger. In this case I'd suggest that you explore other trigger alternatives like simple trigger or daily interval trigger. When using simple trigger as an example you can just configure start time as 00:00 + replicaNumber * 10s and interval as needed (24h or 10min).

So cron is not suitable for your scenario due it obeying the cron format whereas other trigger types allow you this kind of customization. Please note that this setup does not guarantee that there's no locking problems as you job might get delayed or otherwise execute longer than expected (delay between servers). Only way to guarantee such scenario is to either some sort of central processor or locks between nodes (DB locks for example).
